### ISSUE TITLE
ECONNRESET Warning in the Realm Object Server load balancer

### ISSUE THREAD

**Issue by gemiren** [realm/realm-object-server#146]
I am running the realm object server on a Ubuntu 16.04 server. Recently, I've seen a lot of warnings like this one below in the log. The warning will go away if I stop and restart the realm object server. However, the warning comes back quickly. Any ideas or suggestions on how to get rid of this warning? 

The warning message is:
```
[syncProxy] internal error: Error: read ECONNRESET at 
exports._errnoException (util.js:870:11) at TCP.onread (net.js:544:26).
```

Thanks.

**Comment by teotwaki** [comment]
Hi @gemiren,

I've never seen that warning before. Do you have a reproduction case?

Thanks,

**Comment by gemiren** [comment]
I am able to reproduce this on daily basis. Basically, I have to restart the ROS every couple of hours to get rid of this warning and it will come back quickly again :(. When this warning happens, I am not able to sync with the ROS server.

If I look at the ROS status it always says running and the dashboard is displayed on the web browser correctly. In my case it has around 1500 total network connections and current connection is around 250. Realms open is around 1000.

Another thing I noticed is that ROS always runs on a single CPU even though I have multiple CPUs and the other CPUs are ideling. That makes that single CPU goes 100% usage. Is that normal? How can I make the ROS to use all the available CPUs?

Thanks.

**Comment by teotwaki** [comment]
The sync handling is done on a single thread. The dashboard, auth and other services are all on separate threads. The non-developer editions of the ROS allow you to run multiple instances of the sync worker on a single machine (load balancing) to leverage multiple CPUs. This is currently not possible with the developer edition.

Could you provide the full strack trace from the log? It's not possible for us to diagnose much without more information about where the issue comes.

**Comment by gemiren** [comment]
Sure, where can I send the log to? 

Thanks for the CPUs info. That's good to know. I thought the developer version doesn't support load balancing with multiple servers but not CPUs within the same server.

**Comment by teotwaki** [comment]
You can either post the log here (we don't need the full log, just the stack trace under the warning you posted), or send it to help@realm.io.

**Comment by gemiren** [comment]
warn 	2017-03-13 08:46:23.525-0700 	[syncProxy] internal error: Error: read ECONNRESET at exports._errnoException (util.js:870:11) at TCP.onread (net.js:544:26).
debug 	2017-03-13 08:46:23.365-0700 	proxy attempting to upgrade client ::ffff:207.164.79.111:18598 => headers: {"connection":"Upgrade","host":"my_server_ip","sec-websocket-key":"umgbPWnMWYcXI3rKfmKK/A==","sec-websocket-protocol":"io.realm.protocol","sec-websocket-version":"13","upgrade":"websocket","x-realm-access-token":"eyJpZGVudGl0eSI6IjA5OWE0ZjljZmRhYzcwNjY3ZDk1YmVkN2RiN2E2ZmY0IiwicGF0aCI6Ii8wOTlhNGY5Y2ZkYWM3MDY2N2Q5NWJlZDdkYjdhNmZmNC9wcnN5bmNkYiIsImV4cGlyZXMiOjE0ODk0MDcxNDQsImFwcF9pZCI6ImNvbS5pdHVuZXN0b3Bwb2RjYXN0cGxheWVyLmFwcCIsImFjY2VzcyI6WyJkb3dubG9hZCIsInVwbG9hZCIsIm1hbmFnZSJdfQ==:GjtPkbaz2Fi7Vzy67qHbfyw/W0FEeidMEvxuu3u4eQ6LWeYUllDmUeqY9DVMxR3VJP+2V/2XK2AMju8FVYpYZ7G4CvYdh6np3U8D59gbqX/diEBo2wlYJY4YOMHzLJnsUid1vjyBn34/+LfOAGkrVPWB18W42EkKaWzxjNrSyyoD7Ql8Am3sElbA+3n2joRDO8TFj1jzPH6aDb+0DdvNagpKAyplUG1XgMiJsLf4N7GT8YMOyrfX1F91FRmST2V1kFIja53qY+htNBJw5IKy6heJ3URnZoXQz4/W5M+AECTcQv85rY0tDM+YzrRD0XDp8aNpIGxWVM9RZK8QCpITMA==","x-realm-path":"/099a4f9cfdac70667d95bed7db7a6ff4/prsyncdb","x-realm-protocol-version":"16"}.}

**Comment by gemiren** [comment]
I set the log level to all but there is no trace. Only this debug log above.

**Comment by gemiren** [comment]
Another log with a little bit trace in it. Hopefully this helps tracking down the bug. Thanks very much.

```
2017-03-13T18:57:36.444Z - debug: http:  127.0.0.1 - POST /auth HTTP/1.1 200 914 - 5.915 ms
2017-03-13T18:57:36.741Z - warn: [syncProxy] internal error: Error: read ECONNRESET
    at exports._errnoException (util.js:870:11)
    at TCP.onread (net.js:544:26).
2017-03-13T18:57:36.887Z - debug: http:  127.0.0.1 - GET /api/stats HTTP/1.1 304 - - 2.462 ms
2017-03-13T18:57:37.013Z - warn: [syncProxy] internal error: Error: connect ETIMEDOUT 127.0.0.1:27800
    at Object.exports._errnoException (util.js:870:11)
    at exports._exceptionWithHostPort (util.js:893:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1063:14).
```

**Comment by gemiren** [comment]
Is the trace helpful? Any thoughts on what could be issues? I've set the log to trace for a day but that's all the trace I can see in the log. 

I am still seeing this warning all the time. Whenever this warning comes up, the server is not usable any more and needs to be restarted. This makes the realm syncing not usable at all basically. Any help is appreciated. Thanks.

**Comment by morten-krogh** [comment]
@gemiren 

We would like to see the full log.
Can you please send it to mk@realm.io

Thank you so much, and sorry for the trouble.

**Comment by gemiren** [comment]
Hello Morten,

Thanks very much for looking into this issue. Please get the full ROS log from this Dropbox link https://www.dropbox.com/s/995646sg6p974cp/realm-object-server-issue%23146_ECONNRESET_error.log?dl=0. 

The full log has all the logs from last couple of days. The beginning part of the log was set to info level and the later part was set to trace level as told on github.

Hopefully this helps. Please let me know if you need any additional information from me.

Best regards,
Nay

**Comment by gemiren** [comment]
@morten-krogh, thanks very much for looking into this issue. Just sent you the full log through email.

**Comment by bmunkholm** [comment]
@gemiren Very sorry about the delay. It seems like we never received your email. Could you please resend and add this issue number in the subject? Thanks!

**Comment by gemiren** [comment]
Just emailed the full log file to help@realm.io. Thanks.

**Comment by bmunkholm** [comment]
Thanks @gemiren ! We got it this time and are looking into it.

**Comment by morten-krogh** [comment]
@gemiren 

The problem here is that part of the server is stuck because it takes a long time to perform some merge computations.

During that time, the front end of the server keeps operating. 

The ECONNRESET warnings comes because the front end cannot reach the backend of the server.

The real problem is that the server is stuck in a long computation.

We are working on this issue.

I am sorry for the trouble.

**Comment by gemiren** [comment]
Thanks for the information. That's good to know. I think it is normal the backend gets busy and becomes inaccessible to the frontend temporary. However it should be able to recover from that by itself once the long computation is done. What I see so far is that ROS is not able to recover by itself. Once it gives this ECONNRESET warning, it has to be stopped and restarted.

I've updated to ROS 1.3.0 and this issue is still there.

**Comment by morten-krogh** [comment]
It can recover when the computation finishes. The problem is just that it takes so long for the computation to finish that you don't see it. It is better to restart the server in that case.

**Comment by gemiren** [comment]
I had tried couple of times before to leave the ROS to try to recover by itself for 5 to 6 hours but no luck for me. Thanks.

**Comment by dannygrob** [comment]
I have the same issue: internal error:
[syncProxy] internal error: Error: read ECONNRESET at exports._errnoException (util.js:870:11) at TCP.onread (net.js:544:26).

**Comment by bmunkholm** [comment]
@dannygrob @gemiren We are actively working on solving this. I would love to be able to give a timeline, but it's still too early for that as it's not just a quickfix. We appreciate your reports and patience on this.

**Comment by bigfish24** [comment]
This issue was solved in ROS 2.0
